Riot/Modules/Application/LegacyAppDelegate.m

```diff
@@ -2346,38 +2331,84 @@ - (void)handleLaunchAnimation
+        
         if (isLaunching)
         {
-            NSLog(@"[AppDelegate] handleLaunchAnimation: LaunchLoadingView");
-            
-            UIWindow *window = [[UIApplication sharedApplication] keyWindow];
-            
-            if (!launchAnimationContainerView && window)
-            {
-                LaunchLoadingView *launchLoadingView = [LaunchLoadingView instantiate];
-                launchLoadingView.frame = window.bounds;
-                [launchLoadingView updateWithTheme:ThemeService.shared.theme];
-                launchLoadingView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
-                
-                [window addSubview:launchLoadingView];
-                
-                launchAnimationContainerView = launchLoadingView;
-                launchAnimationStart = [NSDate date];
-            }
-            
+            NSLog(@"[AppDelegate] handleAppState: LaunchLoadingView");
+            [self showLaunchAnimation];
             return;
         }
-        else
+
+        [self hideLaunchAnimation];
+        
+        if (self.setPinCoordinatorBridgePresenter)
         {
-            NSLog(@"[AppDelegate] handleLaunchAnimation: isLaunching: NO");
+            NSLog(@"[AppDelegate] handleAppState: PIN code is presented. Do not go further");
+            return;
         }
+        
+        if (mainSession.crypto.crossSigning)
+        {
+            NSLog(@"[AppDelegate] handleAppState: crossSigning.state: %@", @(mainSession.crypto.crossSigning.state));
+            
+            switch (mainSession.crypto.crossSigning.state)
+            {
+                case MXCrossSigningStateCrossSigningExists:
+                    NSLog(@"[AppDelegate] handleAppState: presentVerifyCurrentSessionAlertIfNeededWithSession");
+                    [_masterTabBarController presentVerifyCurrentSessionAlertIfNeededWithSession:mainSession];
+                    break;
+                case MXCrossSigningStateCanCrossSign:
+                    NSLog(@"[AppDelegate] handleAppState: presentReviewUnverifiedSessionsAlertIfNeededWithSession");
+                    [_masterTabBarController presentReviewUnverifiedSessionsAlertIfNeededWithSession:mainSession];
+                    break;
+                default:
+                    break;
+            }
+        }
+        
+        // TODO: We should wait that cross-signing screens are done before going further but it seems fine. Those screens
+        // protect each other.
+        
+        // This is the time to register for incoming requests
+        [self checkPendingRoomKeyRequests];
+        [self checkPendingIncomingKeyVerificationsInSession:mainSession];
+        
+        // Enable listening of incoming key share requests
+        [self enableRoomKeyRequestObserver:mainSession];
+        
+        // Enable listening of incoming key verification requests
+        [self enableIncomingKeyVerificationObserver:mainSession];
```

Comment:
Those methods do not unregister old observers, observer block can be called several times.